indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,694 @@
+      }
+    }
+
+    // Check once again for leadership
+    if (!isLeader.get()) {
+      return;
+    }
+
+    // Process all batches which are due
+    log.debug("Processing all batches which are due for execution. Overall queue size [%d].", processingQueue.size());
+    int numProcessedBatches = 0;
+
+    AllocateRequestBatch nextBatch = processingQueue.peekFirst();
+    while (nextBatch != null && nextBatch.isDue()) {
+      processingQueue.pollFirst();
+      boolean processed;
+      try {
+        processed = processBatch(nextBatch);
+      }
+      catch (Throwable t) {
+        processed = true;
+        log.error(t, "Error while processing batch [%s]", nextBatch.key);
+      }
+
+      ++numProcessedBatches;
+      if (processed) {
+        nextBatch.markCompleted();
+      } else {
+        requeueBatch(nextBatch);
+      }
+
+      nextBatch = processingQueue.peek();
+    }
+
+    // Schedule the next round of processing
+    final long nextScheduleDelay;
+    if (processingQueue.isEmpty()) {
+      nextScheduleDelay = maxWaitTimeMillis;
+    } else {
+      nextBatch = processingQueue.peek();
+      long timeElapsed = System.currentTimeMillis() - nextBatch.getQueueTime();
+      nextScheduleDelay = Math.max(0, maxWaitTimeMillis - timeElapsed);
+    }
+    scheduleQueuePoll(nextScheduleDelay);
+    log.info("Processed [%d] batches, next execution in [%d ms]", numProcessedBatches, nextScheduleDelay);
+  }
+
+  /**
+   * Processes the given batch. Returns true if the batch was completely processed
+   * and should not be requeued.
+   */
+  private boolean processBatch(AllocateRequestBatch requestBatch)
+  {
+    final AllocateRequestKey requestKey = requestBatch.key;
+    keyToBatch.remove(requestKey);
+    if (requestBatch.isEmpty()) {
+      return true;
+    }
+
+    log.debug(
+        "Processing [%d] requests for batch [%s], queue time [%s].",
+        requestBatch.size(),
+        requestKey,
+        requestBatch.getQueueTime()
+    );
+
+    final long startTimeMillis = System.currentTimeMillis();
+    final int batchSize = requestBatch.size();
+    emitBatchMetric("task/action/batch/size", batchSize, requestKey);
+    emitBatchMetric("task/action/batch/queueTime", (startTimeMillis - requestBatch.getQueueTime()), requestKey);
+
+    final Set<DataSegment> usedSegments = retrieveUsedSegments(requestKey);
+    final int successCount = allocateSegmentsForBatch(requestBatch, usedSegments);
+
+    emitBatchMetric("task/action/batch/retries", 1L, requestKey);
+    emitBatchMetric("task/action/batch/runTime", (System.currentTimeMillis() - startTimeMillis), requestKey);
+    log.info("Successfully processed [%d / %d] requests in batch [%s].", successCount, batchSize, requestKey);
+
+    if (requestBatch.isEmpty()) {
+      log.debug("All requests in batch [%s] have been processed.", requestKey);
+      return true;
+    }
+
+    // Requeue the batch only if used segments have changed
```

Comment:
Could you please explain a bit more about this?

Reply:
This is to retain behaviour from the existing flow. If the set of used segments has changed since we started processing the request, it's possible that retrying might make it pass.
https://github.com/apache/druid/blob/79df11c16cd2fd1c6a9867855c42768a079917ff/indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocateAction.java#L223